alignment code not working on pairs

```c++
#include <iostream>

#include <seqan3/alphabet/nucleotide/dna4.hpp>
#include <seqan3/alignment/pairwise/all.hpp>

int main()
{
    using seqan3::operator""_dna4;

    std::pair sequences{"ACCA"_dna4,"ATTA"_dna4};

    seqan3::configuration alignment_config = seqan3::align_cfg::edit |
                                             seqan3::align_cfg::parallel{4};

    seqan3::align_pairwise(sequences, alignment_config);
}
```

Does not compile:

```
/seqan3/test/snippet/test_snippet.cpp: In function ‘int main()’:
/seqan3/test/snippet/test_snippet.cpp:15:55: error: no matching function for call to ‘align_pairwise(std::pair<std::vector<seqan3::dna4>, std::vector<seqan3::dna4> >&, seqan3::configuration<seqan3::align_cfg::mode<seqan3::detail::global_alignment_type>, seqan3::align_cfg::scoring<seqan3::nucleotide_scoring_scheme<signed char> >, seqan3::align_cfg::gap<seqan3::gap_scheme<signed char> >, seqan3::detail::parallel_mode<std::integral_constant<seqan3::detail::align_config_id, seqan3::detail::align_config_id::parallel> > >&)’
   15 |     seqan3::align_pairwise(sequences, alignment_config);
      |                                                       ^
In file included from /seqan3/include/seqan3/alignment/pairwise/all.hpp:15,
                 from /seqan3/test/snippet/test_snippet.cpp:4:
/seqan3/include/seqan3/alignment/pairwise/align_pairwise.hpp:139:16: note: candidate: ‘constexpr auto seqan3::align_pairwise(sequence_t&&, const alignment_config_t&) [with sequence_t = std::pair<std::vector<seqan3::dna4>, std::vector<seqan3::dna4> >&; alignment_config_t = seqan3::configuration<seqan3::align_cfg::mode<seqan3::detail::global_alignment_type>, seqan3::align_cfg::scoring<seqan3::nucleotide_scoring_scheme<signed char> >, seqan3::align_cfg::gap<seqan3::gap_scheme<signed char> >, seqan3::detail::parallel_mode<std::integral_constant<seqan3::detail::align_config_id, seqan3::detail::align_config_id::parallel> > >]’
  139 | constexpr auto align_pairwise(sequence_t && seq, alignment_config_t const & config)
      |                ^~~~~~~~~~~~~~
/seqan3/include/seqan3/alignment/pairwise/align_pairwise.hpp:139:16: note:   constraints not satisfied
In file included from /seqan3/include/seqan3/alignment/pairwise/detail/type_traits.hpp:17,
                 from /seqan3/include/seqan3/alignment/pairwise/align_result_selector.hpp:23,
                 from /seqan3/include/seqan3/alignment/pairwise/alignment_algorithm.hpp:24,
                 from /seqan3/include/seqan3/alignment/pairwise/alignment_configurator.hpp:26,
                 from /seqan3/include/seqan3/alignment/pairwise/align_pairwise.hpp:24,
                 from /seqan3/include/seqan3/alignment/pairwise/all.hpp:15,
                 from /seqan3/test/snippet/test_snippet.cpp:4:
/seqan3/include/seqan3/alignment/pairwise/detail/concept.hpp:109:16: note: within ‘template<class t> concept const bool seqan3::detail::align_pairwise_single_input<t> [with t = std::remove_reference<std::pair<std::vector<seqan3::dna4>, std::vector<seqan3::dna4> >&>::type]’
  109 | SEQAN3_CONCEPT align_pairwise_single_input =
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /seqan3/submodules/range-v3/include/range/v3/algorithm/copy.hpp:25,
                 from /seqan3/include/seqan3/alignment/scoring/scoring_scheme_base.hpp:16,
                 from /seqan3/include/seqan3/alignment/scoring/nucleotide_scoring_scheme.hpp:16,
                 from /seqan3/include/seqan3/alignment/configuration/align_config_edit.hpp:19,
                 from /seqan3/include/seqan3/alignment/configuration/all.hpp:18,
                 from /seqan3/include/seqan3/alignment/pairwise/align_pairwise.hpp:22,
                 from /seqan3/include/seqan3/alignment/pairwise/all.hpp:15,
                 from /seqan3/test/snippet/test_snippet.cpp:4:
/seqan3/submodules/range-v3/include/range/v3/range/concepts.hpp:287:22: note: within ‘template<class T> concept const bool ranges::viewable_range<T> [with T = std::vector<seqan3::dna4>]’
  287 |     CPP_concept_bool viewable_range =
      |                      ^~~~~~~~~~~~~~
/seqan3/submodules/range-v3/include/range/v3/range/concepts.hpp:91:22: note: within ‘template<class T> concept const bool ranges::forwarding_range_<T> [with T = std::vector<seqan3::dna4>]’
   91 |     CPP_concept_bool forwarding_range_ =
      |                      ^~~~~~~~~~~~~~~~~
```

I guess that the main problem here is that `std::pair::first` ( `std::vector<dna4>`) is not a viewable_range, because it is not of type  `std::vector<dna4> &`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -37,6 +37,8 @@ Note that 3.1.0 will be the first API stable release and interfaces in this rele
 
 * The alignment algorithm can now be invoked with a user defined callback function
   ([\#1876](https://github.com/seqan/seqan3/pull/1876)).
+* The function `seqan3::align_pairwise` accepts a std::pair of sequences as input
+  ([\#1913](https://github.com/seqan/seqan3/pull/1913)).
 
 #### Argument Parser
 

--- a/include/seqan3/alignment/pairwise/align_pairwise.hpp
+++ b/include/seqan3/alignment/pairwise/align_pairwise.hpp
@@ -148,6 +148,23 @@ constexpr auto align_pairwise(sequence_t && seq, alignment_config_t const & conf
     return align_pairwise(std::views::single(std::forward<sequence_t>(seq)), config);
 }
 
+//!\overload
+template <typename sequence_t, typename alignment_config_t>
+//!\cond
+    requires tuple_like<sequence_t> && !detail::align_pairwise_single_input<std::remove_reference_t<sequence_t>>
+//!\endcond
+constexpr auto align_pairwise(sequence_t && seq, alignment_config_t const & config)
+{
+    static_assert(std::tuple_size_v<std::remove_reference_t<sequence_t>> == 2,
+                  "Alignment configuration error: Expects exactly two sequences for pairwise alignments.");
+
+    static_assert(std::is_lvalue_reference_v<sequence_t>,
+                  "Alignment configuration error: The given tuple/pair must be an lvalue reference.");
+
+    auto tied_pair = std::tie(get<0>(seq), get<1>(seq));
+    return align_pairwise(tied_pair, config);
+}
+
 //!\cond
 template <typename sequence_t, typename alignment_config_t>
     requires detail::align_pairwise_range_input<sequence_t> &&

--- a/include/seqan3/alignment/pairwise/align_pairwise.hpp
+++ b/include/seqan3/alignment/pairwise/align_pairwise.hpp
@@ -55,27 +55,10 @@ namespace seqan3
  * In cases where only a single alignment is to be computed, the two sequences can be passed as a pair.
  * The pair can be any class template that models the seqan3::tuple_like concept. The tuple elements must model
  * std::ranges::viewable_range and std::copy_constructible.
- * The following examples would compile:
- * ```cpp
- * std::pair p{"ACGTAGC", "AGTACGACG"};
- * auto result = align_pairwise(p, config);
- * ```
- * or you can use std::tie instead as shown in the example below:
- *
- * ```cpp
- * std::vector vec{"ACCA"_dna4, "ATTA"_dna4};
- * using pair_t = decltype(std::tie(vec[0], vec[0]));
- * std::vector<pair_t> source;
- * for (unsigned i = 0; i < vec.size(); ++i)
- * {
- *     for (unsigned j = i + 1; j < vec.size(); ++j)
- *     {
- *         source.push_back(std::tie(vec[i], vec[j]));
- *     }
- * }
- *
- * auto result = seqan3::align_pairwise(source, config);
- * ```
+ * The following example demonstrates how an alignment is computed from a std::pair of sequences.
+ * \snippet test/snippet/alignment/pairwise/align_pairwise.cpp example1
+ * Alternatively, you can use std::tie as shown in the example below:
+ * \snippet test/snippet/alignment/pairwise/align_pairwise.cpp example2
  *
  * ### Compute multiple alignments
  *
@@ -95,7 +78,9 @@ namespace seqan3
  *
  * The following snippets demonstrate the single element and the range based interface.
  *
- * \include test/snippet/alignment/pairwise/align_pairwise.cpp
+ * \snippet test/snippet/alignment/pairwise/align_pairwise.cpp start
+ * \snippet test/snippet/alignment/pairwise/align_pairwise.cpp example3
+ * \snippet test/snippet/alignment/pairwise/align_pairwise.cpp end
  *
  * ### Exception
  *

--- a/include/seqan3/alignment/pairwise/align_pairwise.hpp
+++ b/include/seqan3/alignment/pairwise/align_pairwise.hpp
@@ -55,12 +55,27 @@ namespace seqan3
  * In cases where only a single alignment is to be computed, the two sequences can be passed as a pair.
  * The pair can be any class template that models the seqan3::tuple_like concept. The tuple elements must model
  * std::ranges::viewable_range and std::copy_constructible.
- * Accordingly, the following example wouldn't compile:
+ * The following examples would compile:
  * ```cpp
  * std::pair p{"ACGTAGC", "AGTACGACG"};
- * auto rng = align_pairwise(p, config);
+ * auto result = align_pairwise(p, config);
  * ```
- * You can use std::tie instead as shown in the example below.
+ * or you can use std::tie instead as shown in the example below:
+ *
+ * ```cpp
+ * std::vector vec{"ACCA"_dna4, "ATTA"_dna4};
+ * using pair_t = decltype(std::tie(vec[0], vec[0]));
+ * std::vector<pair_t> source;
+ * for (unsigned i = 0; i < vec.size(); ++i)
+ * {
+ *     for (unsigned j = i + 1; j < vec.size(); ++j)
+ *     {
+ *         source.push_back(std::tie(vec[i], vec[j]));
+ *     }
+ * }
+ *
+ * auto result = seqan3::align_pairwise(source, config);
+ * ```
  *
  * ### Compute multiple alignments
  *

--- a/include/seqan3/alignment/pairwise/align_pairwise.hpp
+++ b/include/seqan3/alignment/pairwise/align_pairwise.hpp
@@ -132,37 +132,29 @@ namespace seqan3
  */
 template <typename sequence_t, typename alignment_config_t>
 //!\cond
-    requires detail::align_pairwise_single_input<std::remove_reference_t<sequence_t>> &&
+    requires detail::align_pairwise_single_input<sequence_t> &&
              std::copy_constructible<std::remove_reference_t<sequence_t>> &&
              detail::is_type_specialisation_of_v<alignment_config_t, configuration>
 //!\endcond
 constexpr auto align_pairwise(sequence_t && seq, alignment_config_t const & config)
 {
-    static_assert(std::tuple_size_v<std::remove_reference_t<sequence_t>> == 2,
-                  "Alignment configuration error: Expects exactly two sequences for pairwise alignments.");
+    using std::get;
 
-    static_assert(std::ranges::viewable_range<std::tuple_element_t<0, std::remove_reference_t<sequence_t>>> &&
-                  std::ranges::viewable_range<std::tuple_element_t<1, std::remove_reference_t<sequence_t>>>,
-                  "Alignment configuration error: The tuple elements must model std::ranges::viewable_range.");
+    if constexpr (std::is_lvalue_reference_v<sequence_t>)  // Forward tuple elements as references.
+    {
+        return align_pairwise(std::tie(get<0>(seq), get<1>(seq)), config);
+    }
+    else
+    {
+        static_assert(std::tuple_size_v<std::remove_reference_t<sequence_t>> == 2,
+                      "Alignment configuration error: Expects exactly two sequences for pairwise alignments.");
 
-    return align_pairwise(std::views::single(std::forward<sequence_t>(seq)), config);
-}
+        static_assert(std::ranges::viewable_range<std::tuple_element_t<0, std::remove_reference_t<sequence_t>>> &&
+                      std::ranges::viewable_range<std::tuple_element_t<1, std::remove_reference_t<sequence_t>>>,
+                      "Alignment configuration error: The tuple elements must model std::ranges::viewable_range.");
 
-//!\overload
-template <typename sequence_t, typename alignment_config_t>
-//!\cond
-    requires tuple_like<sequence_t> && !detail::align_pairwise_single_input<std::remove_reference_t<sequence_t>>
-//!\endcond
-constexpr auto align_pairwise(sequence_t && seq, alignment_config_t const & config)
-{
-    static_assert(std::tuple_size_v<std::remove_reference_t<sequence_t>> == 2,
-                  "Alignment configuration error: Expects exactly two sequences for pairwise alignments.");
-
-    static_assert(std::is_lvalue_reference_v<sequence_t>,
-                  "Alignment configuration error: The given tuple/pair must be an lvalue reference.");
-
-    auto tied_pair = std::tie(get<0>(seq), get<1>(seq));
-    return align_pairwise(tied_pair, config);
+        return align_pairwise(std::views::single(std::forward<sequence_t>(seq)), config);
+    }
 }
 
 //!\cond

--- a/include/seqan3/alignment/pairwise/detail/concept.hpp
+++ b/include/seqan3/alignment/pairwise/detail/concept.hpp
@@ -107,9 +107,10 @@ SEQAN3_CONCEPT indexed_sequence_pair_range = std::ranges::forward_range<t> &&
 //!\cond
 template <typename t>
 SEQAN3_CONCEPT align_pairwise_single_input =
-    sequence_pair<t> &&
-    std::ranges::viewable_range<std::tuple_element_t<0, t>> &&
-    std::ranges::viewable_range<std::tuple_element_t<1, t>>;
+    sequence_pair<std::remove_reference_t<t>> &&
+    std::is_lvalue_reference_v<t> ||
+    (std::ranges::viewable_range<std::tuple_element_t<0, std::remove_reference_t<t>>> &&
+     std::ranges::viewable_range<std::tuple_element_t<1, std::remove_reference_t<t>>>);
 //!\endcond
 
 /*!\interface seqan3::detail::align_pairwise_range_input <>

--- a/test/snippet/alignment/pairwise/align_pairwise.cpp
+++ b/test/snippet/alignment/pairwise/align_pairwise.cpp
@@ -1,6 +1,7 @@
+//![start]
 #include <vector>
 
-#include <seqan3/alignment/configuration/all.hpp>
+#include <seqan3/alignment/configuration/align_config_edit.hpp>
 #include <seqan3/alignment/pairwise/align_pairwise.hpp>
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
 #include <seqan3/core/debug_stream.hpp>
@@ -8,7 +9,26 @@
 int main()
 {
     using seqan3::operator""_dna4;
+//![start]
 
+    // Configure the alignment kernel.
+    auto config = seqan3::align_cfg::edit;
+
+    {
+    //![example1]
+    std::pair p{"ACGTAGC"_dna4, "AGTACGACG"_dna4};
+    auto result = seqan3::align_pairwise(p, config);
+    //![example1]
+    }
+
+    {
+    //![example2]
+    std::vector vec{"ACCA"_dna4, "ATTA"_dna4};
+    auto result = seqan3::align_pairwise(std::tie(vec[0], vec[1]), config);
+    //![example2]
+    }
+
+    //![example3]
     std::vector vec{std::pair{"AGTGCTACG"_dna4, "ACGTGCGACTAG"_dna4},
                     std::pair{"AGTAGACTACG"_dna4, "ACGTACGACACG"_dna4},
                     std::pair{"AGTTACGAC"_dna4, "AGTAGCGATCG"_dna4}};
@@ -20,4 +40,7 @@ int main()
     // Compute the alignment over a range of pairs.
     for (auto const & res : seqan3::align_pairwise(vec, seqan3::align_cfg::edit))
         seqan3::debug_stream << "The score: " << res.score() << "\n";
+    //![example3]
+//![end]
 }
+//![end]

--- a/test/unit/alignment/pairwise/align_pairwise_test.cpp
+++ b/test/unit/alignment/pairwise/align_pairwise_test.cpp
@@ -64,12 +64,12 @@ TYPED_TEST(align_pairwise_test, single_pair)
     auto seq1 = "ACGTGATG"_dna4;
     auto seq2 = "AGTGATACT"_dna4;
 
-    auto p = std::tie(seq1, seq2);
+    auto p1 = std::tie(seq1, seq2);
 
     {  // the score
         seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_score};
 
-        for (auto && res : call_alignment<TypeParam>(p, cfg))
+        for (auto && res : call_alignment<TypeParam>(p1, cfg))
         {
             EXPECT_EQ(res.score(), -4.0);
         }
@@ -78,7 +78,7 @@ TYPED_TEST(align_pairwise_test, single_pair)
     {  // the alignment
         seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_alignment};
         unsigned idx = 0;
-        for (auto && res : call_alignment<TypeParam>(p, cfg))
+        for (auto && res : call_alignment<TypeParam>(p1, cfg))
         {
             EXPECT_EQ(res.id(), idx++);
             EXPECT_EQ(res.score(), -4);
@@ -117,19 +117,15 @@ TYPED_TEST(align_pairwise_test, single_pair)
             EXPECT_EQ(gap2 | seqan3::views::to_char | seqan3::views::to<std::string>, "A-GTGATACT");
         }
     }
-}
 
-TYPED_TEST(align_pairwise_test, single_pair_of_non_lvalues)
-{
-    auto const seq1 = "ACGTGATG"_dna4;
-    auto const seq2 = "AGTGATACT"_dna4;
+    // use make_pair
 
-    auto p = std::make_pair(seq1, seq2);
+    auto p2 = std::make_pair(seq1, seq2);
 
     {  // the score
         seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_score};
 
-        for (auto && res : call_alignment<TypeParam>(p, cfg))
+        for (auto && res : call_alignment<TypeParam>(p2, cfg))
         {
             EXPECT_EQ(res.score(), -4.0);
         }
@@ -138,7 +134,7 @@ TYPED_TEST(align_pairwise_test, single_pair_of_non_lvalues)
     {  // the alignment
         seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_alignment};
         unsigned idx = 0;
-        for (auto && res : call_alignment<TypeParam>(p, cfg))
+        for (auto && res : call_alignment<TypeParam>(p2, cfg))
         {
             EXPECT_EQ(res.id(), idx++);
             EXPECT_EQ(res.score(), -4);

--- a/test/unit/alignment/pairwise/align_pairwise_test.cpp
+++ b/test/unit/alignment/pairwise/align_pairwise_test.cpp
@@ -89,6 +89,38 @@ TYPED_TEST(align_pairwise_test, single_pair)
     }
 }
 
+TYPED_TEST(align_pairwise_test, single_pair_of_non_lvalues)
+{
+    auto const seq1 = "ACGTGATG"_dna4;
+    auto const seq2 = "AGTGATACT"_dna4;
+
+    auto p = std::make_pair(seq1, seq2);
+
+    {  // the score
+        seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_score};
+
+        for (auto && res : call_alignment<TypeParam>(p, cfg))
+        {
+            EXPECT_EQ(res.score(), -4.0);
+        }
+    }
+
+    {  // the alignment
+        seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_alignment};
+        unsigned idx = 0;
+        for (auto && res : call_alignment<TypeParam>(p, cfg))
+        {
+            EXPECT_EQ(res.id(), idx++);
+            EXPECT_EQ(res.score(), -4);
+            EXPECT_EQ(res.back_coordinate().first, 8u);
+            EXPECT_EQ(res.back_coordinate().second, 9u);
+            auto && [gap1, gap2] = res.alignment();
+            EXPECT_EQ(gap1 | seqan3::views::to_char | seqan3::views::to<std::string>, "ACGTGATG--");
+            EXPECT_EQ(gap2 | seqan3::views::to_char | seqan3::views::to<std::string>, "A-GTGATACT");
+        }
+    }
+}
+
 TYPED_TEST(align_pairwise_test, single_pair_double_score)
 {
     if constexpr (!TestFixture::is_vectorised)

--- a/test/unit/alignment/pairwise/align_pairwise_test.cpp
+++ b/test/unit/alignment/pairwise/align_pairwise_test.cpp
@@ -59,11 +59,10 @@ auto call_alignment(seq_t && seq, cfg_t && cfg)
 
 TYPED_TEST(align_pairwise_test, single_pair)
 {
-    // use std::tie
-
     auto seq1 = "ACGTGATG"_dna4;
     auto seq2 = "AGTGATACT"_dna4;
 
+    // use std::tie
     auto p1 = std::tie(seq1, seq2);
 
     {  // the score
@@ -91,7 +90,6 @@ TYPED_TEST(align_pairwise_test, single_pair)
     }
 
     // use std::pair
-
     std::pair sequences{"ACGTGATG"_dna4,"AGTGATACT"_dna4};
 
     {  // the score
@@ -119,7 +117,6 @@ TYPED_TEST(align_pairwise_test, single_pair)
     }
 
     // use make_pair
-
     auto p2 = std::make_pair(seq1, seq2);
 
     {  // the score

--- a/test/unit/alignment/pairwise/align_pairwise_test.cpp
+++ b/test/unit/alignment/pairwise/align_pairwise_test.cpp
@@ -59,6 +59,8 @@ auto call_alignment(seq_t && seq, cfg_t && cfg)
 
 TYPED_TEST(align_pairwise_test, single_pair)
 {
+    // use std::tie
+
     auto seq1 = "ACGTGATG"_dna4;
     auto seq2 = "AGTGATACT"_dna4;
 
@@ -77,6 +79,34 @@ TYPED_TEST(align_pairwise_test, single_pair)
         seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_alignment};
         unsigned idx = 0;
         for (auto && res : call_alignment<TypeParam>(p, cfg))
+        {
+            EXPECT_EQ(res.id(), idx++);
+            EXPECT_EQ(res.score(), -4);
+            EXPECT_EQ(res.back_coordinate().first, 8u);
+            EXPECT_EQ(res.back_coordinate().second, 9u);
+            auto && [gap1, gap2] = res.alignment();
+            EXPECT_EQ(gap1 | seqan3::views::to_char | seqan3::views::to<std::string>, "ACGTGATG--");
+            EXPECT_EQ(gap2 | seqan3::views::to_char | seqan3::views::to<std::string>, "A-GTGATACT");
+        }
+    }
+
+    // use std::pair
+
+    std::pair sequences{"ACGTGATG"_dna4,"AGTGATACT"_dna4};
+
+    {  // the score
+        seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_score};
+
+        for (auto && res : call_alignment<TypeParam>(sequences, cfg))
+        {
+            EXPECT_EQ(res.score(), -4.0);
+        }
+    }
+
+    {  // the alignment
+        seqan3::configuration cfg = seqan3::align_cfg::edit | seqan3::align_cfg::result{seqan3::with_alignment};
+        unsigned idx = 0;
+        for (auto && res : call_alignment<TypeParam>(sequences, cfg))
         {
             EXPECT_EQ(res.id(), idx++);
             EXPECT_EQ(res.score(), -4);

